firebase_test.go

```diff
@@ -338,6 +338,224 @@ func TestVersion(t *testing.T) {
+
+	varName := "GOOGLE_APPLICATION_CREDENTIALS"
+	credOld, credExists := overwriteEnv(varName, "testdata/service_account.json", t)
+	defer reinstateEnv(varName, credOld, credExists)
+
+	_, err := NewApp(context.Background(), &Config{})
+
+	we := "unexpected field databaseUrl in JSON config file"
+	if err == nil || err.Error() != we {
+		t.Errorf("got error = %s; wanted %s", err, we)
+	}
+}
+
+func TestAutoInitBadJson(t *testing.T) {
+	configOld, configExists := overwriteEnv(FirebaseEnvName, "testdata/firebase_config_bad.json", t)
+	defer reinstateEnv(FirebaseEnvName, configOld, configExists)
+
+	varName := "GOOGLE_APPLICATION_CREDENTIALS"
+	credOld, credExists := overwriteEnv(varName, "testdata/service_account.json", t)
+	defer reinstateEnv(varName, credOld, credExists)
+
+	_, err := NewApp(context.Background(), &Config{})
+
+	we := "invalid character 'b' looking for beginning of value"
+	if err == nil || err.Error() != we {
+		t.Errorf("got error = %s; wanted %s", err, we)
+	}
+}
+
+func TestAutoInitEmptyJson(t *testing.T) {
+	configOld, configExists := overwriteEnv(FirebaseEnvName, "testdata/firebase_config_empty.json", t)
+	defer reinstateEnv(FirebaseEnvName, configOld, configExists)
+
+	varName := "GOOGLE_APPLICATION_CREDENTIALS"
+	credOld, credExists := overwriteEnv(varName, "testdata/service_account.json", t)
+	defer reinstateEnv(varName, credOld, credExists)
+
+	_, err := NewApp(context.Background(), &Config{})
+
+	we := "unexpected end of JSON input"
+	if err == nil || err.Error() != we {
+		t.Errorf("got error = %s; wanted %s", err, we)
+	}
+}
+
+func TestAutoInitNilOptionsNoConfig(t *testing.T) {
+	configOld, configExists := overwriteEnv(FirebaseEnvName, "", t)
+	defer reinstateEnv(FirebaseEnvName, configOld, configExists)
+	os.Unsetenv(FirebaseEnvName)
+
+	varName := "GOOGLE_APPLICATION_CREDENTIALS"
+	credOld, credExists := overwriteEnv(varName, "testdata/service_account.json", t)
+	defer reinstateEnv(varName, credOld, credExists)
+
+	app, err := NewApp(context.Background(), nil)
+	if err != nil {
+		t.Errorf("got error = %s; wanted nil", err)
+	}
+
+	want := &Config{
+		DatabaseURL:   "",
+		ProjectID:     "mock-project-id", // from default credentials
+		StorageBucket: "",
+	}
+	compareConfig(app, want, t)
+}
+
+func TestAutoInitNilOptionsWithConfig(t *testing.T) {
+	configOld, configExists := overwriteEnv(FirebaseEnvName, "testdata/firebase_config.json", t)
+	defer reinstateEnv(FirebaseEnvName, configOld, configExists)
+
+	varName := "GOOGLE_APPLICATION_CREDENTIALS"
+	credOld, credExists := overwriteEnv(varName, "testdata/service_account.json", t)
+	defer reinstateEnv(varName, credOld, credExists)
+
+	app, err := NewApp(context.Background(), nil)
+	if err != nil {
+		t.Errorf("got error = %s; wanted nil", err)
+	}
+
+	want := &Config{
+		DatabaseURL:   "https://hipster-chat.firebaseio.mock",
+		ProjectID:     "hipster-chat-mock",
+		StorageBucket: "hipster-chat.appspot.mock",
+	}
+	compareConfig(app, want, t)
```

Comment:
These tests all have the same basic structure, and I wonder if it wouldn't be clearer to do this as a table-driven test.
https://blog.golang.org/subtests

The test could be defined as a:

    struct {
        env     string
        initial &Config
        want    &Config
    }

Then the one test function could grab the env, defer reinstating it, and then then loop over each case setting the config, calling NewApp, and then comparing the result. I'm curious what @hiranya911 thinks.

Reply:
Yeah, I did a similar thing in the python tests, I agree it is cleaner.

Reply:
+1